services/src/main/java/com/google/bos/iot/core/bambi/BambiService.java

```diff
@@ -0,0 +1,412 @@
+    SheetsAppender.setSheetsOutputStream(stream);
+    try {
+      action.run();
+    } catch (Exception e) {
+      LOGGER.error("Exception during processing request: {}", e.getMessage(), e);
+    } finally {
+      LOGGER.info("Finished processing request for this sheet.");
+      stream.appendToSheet();
+      SheetsAppender.setSheetsOutputStream(null);
+    }
+  }
+
+  private void processRequest(BambiRequestParams params, String spreadsheetId, String date) {
+    String repoDir = Paths.get(baseCloningDir, params.registryId).toString();
+    try {
+      prepareDirectory(repoDir);
+      try (GenericGitRepository repository = getRepository(params.registryId, repoDir)) {
+        repository.cloneRepo(params.importBranch);
+        handleTask(params, repository, spreadsheetId, date);
+      }
+    } catch (Exception e) {
+      LOGGER.error("Could not complete git operation, failed with error: {}", e.getMessage(), e);
+    }
+  }
+
+  private void handleTask(BambiRequestParams params, GenericGitRepository repository,
+      String spreadsheetId, String date) throws Exception {
+    String udmiModelPath = Paths.get(repository.getDirectory(), "udmi").toString();
+    if (IMPORT_REQUEST.equals(params.requestType)) {
+      handleImport(spreadsheetId, udmiModelPath);
+    } else if (EXPORT_REQUEST.equals(params.requestType)) {
+      handleExport(spreadsheetId, udmiModelPath, repository,
+          String.format("proposal.%s.%s", spreadsheetId, date));
+    }
+  }
+
+  private void handleImport(String spreadsheetId, String udmiModelPath) {
+    LOGGER.info("Populating site model in bambi sheet from {}", udmiModelPath);
+    new BambiSync(spreadsheetId, udmiModelPath).execute();
+  }
+
+  private void handleExport(String spreadsheetId, String udmiModelPath,
+      GenericGitRepository repository, String exportBranch) throws Exception {
+    LOGGER.info("Merging data from BAMBI to site model on disk at {}", udmiModelPath);
+    repository.createAndCheckoutBranch(exportBranch);
+    new LocalDiskSync(spreadsheetId, udmiModelPath).execute();
+
+    if (repository.isWorkingTreeClean()) {
+      LOGGER.info("Working tree is clean, nothing to commit.");
+    } else {
+      LOGGER.info("Committing and pushing changes to branch {}", exportBranch);
+      repository.add(".");
+      repository.commit("Merge changes from BAMBI");
+      repository.push();
+      LOGGER.info("Push successful.");
+    }
+  }
+
+  private Optional<String> getSpreadsheetId(String sourceUrl) {
+    Matcher matcher = SPREADSHEET_ID_PATTERN.matcher(sourceUrl);
+    return matcher.find() ? Optional.of(matcher.group(1)) : Optional.empty();
+  }
+
+  private void prepareDirectory(String directory) throws IOException {
+    File dirFile = new File(directory);
+    if (dirFile.exists()) {
+      LOGGER.info("Deleting existing directory {}", directory);
+      deleteDirectory(dirFile);
+    }
+  }
+
+  private GenericGitRepository getRepository(String repoName, String repoClonePath)
```

Comment:
This should be pushed down into the `Repository` abstraction, since it's not really a function of Bambi specifically.